crates/lang_srv/src/server.rs

```diff
@@ -61,34 +68,123 @@ impl RocLs {
             semantic_tokens_provider: Some(semantic_tokens_provider),
+            completion_provider: Some(completion_provider),
             ..ServerCapabilities::default()
         }
     }
 
     /// Records a document content change.
     async fn change(&self, fi: Url, text: String, version: i32) {
-        self.registry()
-            .apply_change(DocumentChange::Modified(fi.clone(), text));
+        let updating_result = self.inner.change(&fi, text, version).await;
 
-        let diagnostics = match std::panic::catch_unwind(|| self.registry().diagnostics(&fi)) {
-            Ok(ds) => ds,
-            Err(_) => return,
-        };
+        //The analysis task can be cancelled by another change coming in which will update the watched variable
+        if let Err(e) = updating_result {
+            debug!("cancelled change. Reason:{:?}", e);
+            return;
+        }
+        debug!("applied_change getting and returning diagnostics");
+
+        let diagnostics = self.inner.registry.diagnostics(&fi).await;
 
         self.client
             .publish_diagnostics(fi, diagnostics, Some(version))
             .await;
     }
+}
+
+impl Inner {
+    pub fn new() -> Inner {
+        Self {
+            registry: Registry::default(),
+        }
+    }
+
+    async fn registry(&self) -> &Registry {
+        &self.registry
+    }
+
+    async fn close(&self, _fi: Url) {
+        ()
+    }
+
+    pub async fn change(
+        &self,
+        fi: &Url,
+        text: String,
+        version: i32,
+    ) -> std::result::Result<(), String> {
+        debug!("V{:?}:starting change", version);
+        //was write lock
+
+        debug!("V{:?}:change acquired registry lock", version);
+        let (results, partial) = global_analysis(fi.clone(), text, version);
+
+        self.registry
+            .apply_doc_info_changes(fi.clone(), partial.clone())
+            .await;
+        //Now that we've got our new partial document written and we hold the exclusive write_handle to its analysis we can allow other tasks to access the registry and the doc_info inside this partial document
+
+        debug!(
+            "V{:?}:finished updating docinfo, starting analysis ",
+            version
+        );
+
+        let inner_ref = self;
+        let updating_result = async {
+            //This reduces wasted computation by waiting to see if a new change comes in, but does delay the final analysis. Ideally this would be replaced with cancelling the analysis when a new one comes in.
+            tokio::time::sleep(Duration::from_millis(100)).await;
+            let is_latest = inner_ref
+                .registry
+                .get_latest_version(fi)
+                .await
+                .map(|latest| latest == version)
+                .unwrap_or(true);
+            if !is_latest {
+                return Err("Not latest version skipping analysis".to_string());
+            }
```

Comment:
Have you tested if this check is necessary for a good experience? I would honestly be okay with just removing it for now. I don't think the 100ms sleep at the head helps, because all changes would hit that sleep, so it only delays the time at which the chain of analyses occur (+/- some jitter/non-determinism in the event loop it induces, of course)

Reply:
I have and it makes a significant difference on slow hardware for me. 
Because we aren't actually cancelling the analysis task, just discarding the result, when I'm typing many of these tasks stack up in the background and slow down the computer. It actually makes the final analysis take much longer on my slower laptop. 
If i add this slight delay before checking the version it gives time for the next change event to update the docinfo so that when we check the latest version it doesn't match and we don't start the analysis at all for most changes coming in during typing.

Reply:
Ahhh that's a great point. I forgot that the doc info is updated before this.